### PR TITLE
feat(token): refresh token before a request made if digest fails

### PR DESCRIPTION
**Pull request by reinos**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [our conventions](../CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently we have a digest in place (`setTimeout`) that will execute when the token expired to fetch a new one.

Issue Number: N/A

## What is the new behavior?
To be **100%** sure that we have a valid token, we do a simple check before a request is made. If the token got expired, fetch a new token just before the actual request can be made and use that one instead.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/api/core/componentStorage.ts [hunk 1]
```diff
@@ -1,6 +1,19 @@
 // tslint:disable:max-classes-per-file
 import { Tokens } from "./tokenManager";
 
+/**
+ * fetch an arrau of all aliases for one access token
```
Comment (veger) [inline src/api/core/componentStorage.ts:5]:
```suggestion
 * fetch an array of all aliases for one access token
```

File: src/api/core/componentStorage.ts [hunk 2]
```diff
@@ -1,6 +1,19 @@
...
+ * @internal
+ */
+export const getAliases = (accessToken: string, tokens: tokenList): string[] => Object.entries(tokens)
```
Comment (veger) [inline src/api/core/componentStorage.ts:8]: What is an 'alias'? When I look at this function it looks like you just fetch the 'full' token from the list of available tokens using the provided accessToken?

File: src/api/core/tokenManager.ts [hunk 3]
```diff
@@ -201,18 +218,33 @@ export class TokenManager {
...
       this.logger.debug("tokenManager", `refresh ${aliases[0]} token`);
-      this.refreshes.delete(accessToken);
       this.refresh(aliases)
+        .pipe(
+          finalize(() => this.removeRefreshDigest(accessToken)),
+        )
         .subscribe(
           ({ access_token }: Tokens) => this.startRefreshDigest(aliases, access_token), // start a digest for the new token
-          () => this.terminate() // TODO: send an event out, so the user can act on that
+          () => this.terminate(), // TODO: send an event out, so the user can act on that,
```
Comment (veger) [inline src/api/core/tokenManager.ts:227]:
```suggestion
          () => this.terminate(), // TODO: send an event out, so the user can act on that
```

**Review by veger — APPROVED** [review]

File: src/api/core/componentStorage.ts (hunk 2 above)
Comment (reinos) [reply src/api/core/componentStorage.ts]: You can set multiple aliases per token as you can see here --> https://github.com/jexia/jexia-sdk-js/blob/master/src/api/core/tokenManager.ts#L184-L194

As you can see, there is a filter that fetch an alias names that match the access_token. 

```
.filter(([, token]) => token.access_token === accessToken)
```

See also https://docs.jexia.com/ums/#sign-in-a-user for the use of alias ;-)

**Pull request merged**